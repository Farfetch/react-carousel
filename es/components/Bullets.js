function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import Styles from "./Bullets.css";
import cx from 'classnames';

var Bullets = /*#__PURE__*/function (_PureComponent) {
  _inherits(Bullets, _PureComponent);

  var _super = _createSuper(Bullets);

  _createClass(Bullets, null, [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      var bullets = Math.ceil(props.itemsLength / props.itemsToScroll);
      var activeBullet = Math.ceil(props.activeItem / props.itemsToScroll);
      return {
        isInfinite: bullets > state.maxBullets,
        translationValue: props.isRTL ? 1.8 : -1.8,
        itemsToScroll: props.itemsToScroll,
        bullets: bullets,
        activeBullet: activeBullet
      };
    }
  }]);

  function Bullets(props) {
    var _this;

    _classCallCheck(this, Bullets);

    _this = _super.call(this, props);
    _this.state = {
      maxBullets: 5,
      translationValue: 0,
      isInfinite: false,
      itemsToScroll: 1,
      bullets: 5,
      activeBullet: 1
    };
    return _this;
  }

  _createClass(Bullets, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          activeItem = _this$props.activeItem,
          itemsLength = _this$props.itemsLength,
          theme = _this$props.theme,
          isRTL = _this$props.isRTL,
          itemsToScroll = _this$props.itemsToScroll,
          otherProps = _objectWithoutProperties(_this$props, ["activeItem", "itemsLength", "theme", "isRTL", "itemsToScroll"]);

      var isInfinite = this.state.isInfinite;
      var cxs = this.getClassNamesToApply();
      return /*#__PURE__*/React.createElement("div", _extends({
        className: cxs,
        "aria-hidden": "true",
        "data-active": activeItem
      }, otherProps), isInfinite ? /*#__PURE__*/React.createElement("div", {
        className: Styles.moveInfinite,
        style: {
          transform: "translate(".concat(this.getTranslationValue(), "rem)")
        }
      }, this.getBullets()) : this.getBullets());
    }
  }, {
    key: "getSecondaryBullets",
    value: function getSecondaryBullets() {
      var _this$state = this.state,
          activeBullet = _this$state.activeBullet,
          bullets = _this$state.bullets;
      return {
        // Secondary bullet is the previous one, unless active bullet is the first
        first: activeBullet === 0 ? 2 : activeBullet - 1,
        // Other secondary bullet is the next one, unless active bullet is the last one
        second: activeBullet === bullets ? activeBullet - 2 : activeBullet + 1
      };
    }
  }, {
    key: "getTranslationValue",
    value: function getTranslationValue() {
      var _this$state2 = this.state,
          translationValue = _this$state2.translationValue,
          maxBullets = _this$state2.maxBullets,
          bullets = _this$state2.bullets,
          activeBullet = _this$state2.activeBullet; // If active slide is in first, second or third position translation doesn't happen

      if (bullets <= maxBullets || activeBullet <= 2) {
        return 0;
      } // If active slide is one of the last three positions translation stays the same


      if (activeBullet > bullets - 3) {
        return (bullets - maxBullets) * translationValue;
      } // Generic case where the bullets are translated


      return (activeBullet - 2) * translationValue;
    }
  }, {
    key: "getBullets",
    value: function getBullets() {
      var _this$props2 = this.props,
          activeItem = _this$props2.activeItem,
          theme = _this$props2.theme;
      var _this$state3 = this.state,
          isInfinite = _this$state3.isInfinite,
          activeBullet = _this$state3.activeBullet,
          bullets = _this$state3.bullets;
      var secondaryBullets = this.getSecondaryBullets();
      var itemsNodes = [];

      var getCxInfinite = function getCxInfinite(index) {
        var _cx;

        return cx(Styles.bulletInfinite, (_cx = {}, _defineProperty(_cx, Styles.isActive, activeBullet === index), _defineProperty(_cx, Styles.isSecondary, secondaryBullets.first === index || secondaryBullets.second === index), _defineProperty(_cx, theme.isActive, theme.isActive && activeItem === index), _cx));
      };

      var getCxDefault = function getCxDefault(index) {
        var _cx2;

        return cx(Styles.bullet, (_cx2 = {}, _defineProperty(_cx2, Styles.isActive, activeBullet === index), _defineProperty(_cx2, theme.isActive, theme.isActive && activeItem === index), _cx2));
      };

      var generateClass = isInfinite ? getCxInfinite : getCxDefault;

      for (var i = 0; i < bullets; i++) {
        itemsNodes.push( /*#__PURE__*/React.createElement("span", {
          key: i,
          className: generateClass(i),
          "data-tstid": "bullet"
        }));
      }

      return itemsNodes;
    }
  }, {
    key: "getClassNamesToApply",
    value: function getClassNamesToApply() {
      var theme = this.props.theme;
      var isInfinite = this.state.isInfinite;

      var baseThemeClasses = _defineProperty({}, theme.container, !!theme.container);

      if (isInfinite) {
        return cx(Styles.containerInfinite, _objectSpread(_objectSpread({}, baseThemeClasses), {}, _defineProperty({}, theme.containerInfinite, !!theme.containerInfinite)));
      }

      return cx(Styles.containerDefault, _objectSpread(_objectSpread({}, baseThemeClasses), {}, _defineProperty({}, theme.containerDefault, !!theme.containerDefault)));
    }
  }]);

  return Bullets;
}(PureComponent);

Bullets.propTypes = {
  /* Index of active item - start on 0 */
  activeItem: PropTypes.number.isRequired,

  /* Items length */
  itemsLength: PropTypes.number.isRequired,

  /* Navigation direction  */
  isRTL: PropTypes.bool,

  /* SliderNav theme */
  theme: PropTypes.shape({
    container: PropTypes.string,
    containerDefault: PropTypes.string,
    containerInfinite: PropTypes.string,
    moveInfinite: PropTypes.string,
    bullet: PropTypes.string,
    bulletInfinite: PropTypes.string,
    isSecondary: PropTypes.string,
    isActive: PropTypes.string
  }),
  itemsToScroll: PropTypes.number
};
Bullets.defaultProps = {
  theme: {},
  itemsToScroll: 1
};
export default Bullets;