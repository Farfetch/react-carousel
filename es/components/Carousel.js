function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

import { CarouselProvider } from "../context";
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import Styles from "./Carousel.css";
import cx from 'classnames';

var Carousel = /*#__PURE__*/function (_Component) {
  _inherits(Carousel, _Component);

  var _super = _createSuper(Carousel);

  function Carousel() {
    _classCallCheck(this, Carousel);

    return _super.apply(this, arguments);
  }

  _createClass(Carousel, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          children = _this$props.children,
          isInfinite = _this$props.isInfinite,
          isRTL = _this$props.isRTL,
          itemsToShow = _this$props.itemsToShow,
          itemsToScroll = _this$props.itemsToScroll,
          onAfterChange = _this$props.onAfterChange,
          startItem = _this$props.startItem,
          otherProps = _objectWithoutProperties(_this$props, ["className", "children", "isInfinite", "isRTL", "itemsToShow", "itemsToScroll", "onAfterChange", "startItem"]);

      var providerProps = {
        isInfinite: isInfinite,
        isRTL: isRTL,
        itemsToShow: itemsToShow,
        onAfterChange: onAfterChange,
        itemsToScroll: itemsToScroll,
        startItem: startItem
      };
      return /*#__PURE__*/React.createElement(CarouselProvider, providerProps, /*#__PURE__*/React.createElement("div", _extends({
        className: cx(Styles.container, className)
      }, otherProps), children));
    }
  }]);

  return Carousel;
}(Component);

Carousel.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  isInfinite: PropTypes.bool,
  isRTL: PropTypes.bool,
  itemsToShow: PropTypes.number,
  onAfterChange: PropTypes.func,
  itemsToScroll: PropTypes.number,
  startItem: PropTypes.number
};
Carousel.defaultProps = {
  itemsToScroll: 1
};
export default Carousel;