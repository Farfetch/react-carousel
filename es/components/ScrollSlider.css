.container {
    overflow-x: scroll;
    overflow-y: hidden;

    /* Hide scrollbar on Chrome and Safari */

    &::-webkit-scrollbar {
        width: 0;
        display: none;
    }

    /* Hide scrollbar on Firefox */
    scrollbar-width: none;

    /* Hide scrollbar on IE */
    -ms-overflow-style: none;
}
