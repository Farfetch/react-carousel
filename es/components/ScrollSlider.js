function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

import { animateScroll, renderSlides } from "../utils";
import PropTypes from 'prop-types';
import React, { Children, Component, createRef } from 'react';
import ScrollStyles from "./ScrollSlider.css";
import Styles from "./Slider.css";
import cx from 'classnames';
import debounce from 'lodash/debounce';

var ScrollSlider = /*#__PURE__*/function (_Component) {
  _inherits(ScrollSlider, _Component);

  var _super = _createSuper(ScrollSlider);

  function ScrollSlider(props) {
    var _this;

    _classCallCheck(this, ScrollSlider);

    _this = _super.call(this, props);
    _this._verifyActiveItemDebounce = debounce(_this._verifyActiveItem, 25);
    _this._handleScroll = _this._handleScroll.bind(_assertThisInitialized(_this));
    _this.itemsLength = Children.count(props.children);
    _this.activeItem = 0;
    _this.shouldMoveToItem = false;
    _this.isPreventingScroll = false;
    _this.containerRef = /*#__PURE__*/createRef();
    return _this;
  }

  _createClass(ScrollSlider, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.props.setItemsLength(this.itemsLength);
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      if (this.shouldMoveToItem) {
        this._moveToItem();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          activeItem = _this$props.activeItem,
          className = _this$props.className,
          children = _this$props.children,
          goTo = _this$props.goTo,
          isRTL = _this$props.isRTL,
          itemsToShow = _this$props.itemsToShow,
          limitScroll = _this$props.limitScroll,
          setItemsLength = _this$props.setItemsLength,
          othersProps = _objectWithoutProperties(_this$props, ["activeItem", "className", "children", "goTo", "isRTL", "itemsToShow", "limitScroll", "setItemsLength"]);

      if (activeItem !== this.activeItem) {
        this.activeItem = activeItem;
        this.shouldMoveToItem = true;
      } else {
        this.shouldMoveToItem = false;
      }

      var style = {
        width: "".concat((100 / itemsToShow).toFixed(6), "%")
      };
      return /*#__PURE__*/React.createElement("div", _extends({
        ref: this.containerRef,
        className: cx(Styles.slider, ScrollStyles.container, className),
        onScroll: this._handleScroll
      }, othersProps), renderSlides(children, itemsToShow, style));
    }
  }, {
    key: "_moveToItem",
    value: function _moveToItem() {
      var _this2 = this;

      var el = this.containerRef.current;

      if (!el) {
        return null;
      }

      var offsetWidth = el.offsetWidth;
      var _this$props2 = this.props,
          activeItem = _this$props2.activeItem,
          itemsToShow = _this$props2.itemsToShow,
          isRTL = _this$props2.isRTL;
      var itemToMoveTo = isRTL ? this.itemsLength - 2 - activeItem : activeItem;
      var finalPosition = offsetWidth / itemsToShow * itemToMoveTo;
      animateScroll(el, finalPosition, 0.3, function () {
        _this2.shouldMoveToItem = false;
      });
    }
    /* Don't work on Chrome, but work on iOS */

  }, {
    key: "_forceScrollSnapStop",
    value: function _forceScrollSnapStop(newPosition, newIndex) {
      var _this3 = this;

      var el = this.containerRef.current;

      if (!el) {
        return null;
      }

      el.scrollLeft = newPosition;
      this.isPreventingScroll = true;
      el.style.overflow = 'hidden';
      setTimeout(function () {
        _this3.isPreventingScroll = false;
        el.style.overflow = '';
      }, 80);
      this.props.goTo(newIndex);
    }
  }, {
    key: "_verifyActiveItem",
    value: function _verifyActiveItem() {
      var _this$props3 = this.props,
          goTo = _this$props3.goTo,
          itemsToShow = _this$props3.itemsToShow,
          isRTL = _this$props3.isRTL;
      var _this$containerRef$cu = this.containerRef.current,
          offsetWidth = _this$containerRef$cu.offsetWidth,
          scrollLeft = _this$containerRef$cu.scrollLeft;
      var activeFraction = Math.round(scrollLeft / offsetWidth * 100) / 100;
      var newActiveItem = isRTL ? this.itemsLength - 1 - Math.round(activeFraction * itemsToShow) - (itemsToShow - 1) : Math.round(activeFraction * itemsToShow);

      if (this.activeItem === newActiveItem) {
        return null;
      }

      this.activeItem = newActiveItem;
      goTo(newActiveItem);
    }
  }, {
    key: "_handleScroll",
    value: function _handleScroll(e) {
      if (this.isPreventingScroll) {
        e.stopPropagation();
        e.preventDefault();
        return null;
      }

      if (this.shouldMoveToItem) {
        return null;
      }

      var _this$props4 = this.props,
          limitScroll = _this$props4.limitScroll,
          itemsToShow = _this$props4.itemsToShow;

      if (limitScroll) {
        var _this$containerRef$cu2 = this.containerRef.current,
            offsetWidth = _this$containerRef$cu2.offsetWidth,
            scrollLeft = _this$containerRef$cu2.scrollLeft; // When scrolling too fast, we want to prevent scrolling 2 slides at once...

        var newPrevIndex = this.activeItem - 1;
        var newNextIndex = this.activeItem + 1;
        var limitScrollNext = offsetWidth / itemsToShow * newNextIndex;
        var limitScrollPrev = offsetWidth / itemsToShow * newPrevIndex; // ...so, we need to force a scroll stop limit position when that happens ...

        if (scrollLeft > limitScrollNext) {
          return this._forceScrollSnapStop(limitScrollNext, newNextIndex);
        } else if (scrollLeft < limitScrollPrev) {
          return this._forceScrollSnapStop(limitScrollPrev, newPrevIndex);
        }
      }

      this._verifyActiveItemDebounce();
    }
  }]);

  return ScrollSlider;
}(Component);

ScrollSlider.propTypes = {
  activeItem: PropTypes.number.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  isRTL: PropTypes.bool.isRequired,
  itemsToShow: PropTypes.number.isRequired,
  setItemsLength: PropTypes.func.isRequired,
  goTo: PropTypes.func.isRequired,
  limitScroll: PropTypes.bool
};
export default ScrollSlider;