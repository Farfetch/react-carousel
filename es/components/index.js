export { default as Arrow } from "./Arrow";
export { default as Bullets } from "./Bullets";
export { default as Carousel } from "./Carousel";
export { default as ScrollSlider } from "./ScrollSlider";
export { default as UnevenItemsScrollSlider } from "./UnevenItemsScrollSlider";
export { default as SnapSlider } from "./SnapSlider";
export { default as SwipeSlider } from "./SwipeSlider";