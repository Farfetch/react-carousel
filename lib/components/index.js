"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Arrow", {
  enumerable: true,
  get: function get() {
    return _Arrow["default"];
  }
});
Object.defineProperty(exports, "Bullets", {
  enumerable: true,
  get: function get() {
    return _Bullets["default"];
  }
});
Object.defineProperty(exports, "Carousel", {
  enumerable: true,
  get: function get() {
    return _Carousel["default"];
  }
});
Object.defineProperty(exports, "ScrollSlider", {
  enumerable: true,
  get: function get() {
    return _ScrollSlider["default"];
  }
});
Object.defineProperty(exports, "UnevenItemsScrollSlider", {
  enumerable: true,
  get: function get() {
    return _UnevenItemsScrollSlider["default"];
  }
});
Object.defineProperty(exports, "SnapSlider", {
  enumerable: true,
  get: function get() {
    return _SnapSlider["default"];
  }
});
Object.defineProperty(exports, "SwipeSlider", {
  enumerable: true,
  get: function get() {
    return _SwipeSlider["default"];
  }
});

var _Arrow = _interopRequireDefault(require("./Arrow"));

var _Bullets = _interopRequireDefault(require("./Bullets"));

var _Carousel = _interopRequireDefault(require("./Carousel"));

var _ScrollSlider = _interopRequireDefault(require("./ScrollSlider"));

var _UnevenItemsScrollSlider = _interopRequireDefault(require("./UnevenItemsScrollSlider"));

var _SnapSlider = _interopRequireDefault(require("./SnapSlider"));

var _SwipeSlider = _interopRequireDefault(require("./SwipeSlider"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }