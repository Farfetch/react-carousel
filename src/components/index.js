export { default as Arrow } from './Arrow';
export { default as Bullet } from './Bullet';
export { default as Bullets } from './Bullets';
export { default as BulletsInfiniteWrapper } from './BulletsInfiniteWrapper';
export { default as Carousel } from './Carousel';
export { default as SnapSlider } from './SnapSlider';
export { default as ScrollSlider } from './ScrollSlider';
export { default as UnevenItemsScrollSlider } from './UnevenItemsScrollSlider';
export { default as SwipeSlider } from './SwipeSlider';
