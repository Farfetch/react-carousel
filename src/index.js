export { checkSnapSupport } from '#utils';
export { CarouselContext, CarouselProvider } from '#context';
export { Arrow, BaseArrow, Carousel, SnapSlider } from '#components';
export {
    Bullets,
    SwipeSlider,
    ScrollSlider,
    UnevenItemsScrollSlider,
} from '#containers';
