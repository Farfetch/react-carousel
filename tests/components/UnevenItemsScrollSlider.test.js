import { createSerializer } from 'enzyme-to-json';
// At the time of writing, enzyme still does not support shallow for useEffect
import { UnevenItemsScrollSlider } from '../../src/components';
import { shallow } from 'enzyme';
import React from 'react';

const children = [
    <div key="0">First slide</div>,
    <div key="1">Second slide</div>,
    <div key="2">Third slide</div>,
    <div key="3">Fourth slide</div>,
];

const mockRequiredProps = {
    activeItem: 0,
    children,
    isRTL: false,
    itemsLength: 5,
    setItemsLength: jest.fn(),
    goTo: jest.fn(),
};

describe('<UnevenItemsScrollSlider />', () => {
    beforeAll(() => {
        jest.useFakeTimers();
        expect.addSnapshotSerializer(createSerializer());
    });

    afterEach(() => {
        jest.clearAllMocks();
        jest.clearAllTimers();
    });

    describe('<UnevenItemsScrollSlider /> - props', () => {
        it('should render correctly with the required props', () => {
            const wrapper = shallow(
                <UnevenItemsScrollSlider {...mockRequiredProps} />
            );

            expect(wrapper).toMatchSnapshot();
        });
    });
});
