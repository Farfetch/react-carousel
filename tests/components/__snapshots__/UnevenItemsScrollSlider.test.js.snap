// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UnevenItemsScrollSlider /> <UnevenItemsScrollSlider /> - props should render correctly with additional props 1`] = `
<div
  className="slider dummyClassName scrollSliderContainer"
  onScroll={[Function]}
>
  <div
    className="unevenItemsWrapper"
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="slide"
      key=".$0"
      style={Object {}}
    >
      First slide
    </div>
    <div
      className="slide"
      key=".$1"
      style={Object {}}
    >
      Second slide
    </div>
    <div
      className="slide"
      key=".$2"
      style={Object {}}
    >
      Third slide
    </div>
    <div
      className="slide"
      key=".$3"
      style={Object {}}
    >
      Fourth slide
    </div>
  </div>
</div>
`;

exports[`<UnevenItemsScrollSlider /> <UnevenItemsScrollSlider /> - props should render correctly with the required props 1`] = `
<div
  className="slider scrollSliderContainer"
  onScroll={[Function]}
>
  <div
    className="unevenItemsWrapper"
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="slide"
      key=".$0"
      style={Object {}}
    >
      First slide
    </div>
    <div
      className="slide"
      key=".$1"
      style={Object {}}
    >
      Second slide
    </div>
    <div
      className="slide"
      key=".$2"
      style={Object {}}
    >
      Third slide
    </div>
    <div
      className="slide"
      key=".$3"
      style={Object {}}
    >
      Fourth slide
    </div>
  </div>
</div>
`;
